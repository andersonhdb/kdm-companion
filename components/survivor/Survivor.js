import React from 'react'
import {
  View,
  Text,
  Title,
  Subtitle,
  Image,
  Icon,
  Button,
  Row,
  Caption,
  Divider,
} from '@shoutem/ui'

import { observer, inject } from 'mobx-react/native'
import PropTypes from 'prop-types'
import R from 'ramda'

import GenderButton from './GenderButton'
import EditableProperty from './EditableProperty'
import FightingArts from './FightingArts'
import Disorders from './Disorders'
import Note from '../Note'

const ico_movement = require('../../images/icon_movement-24.png')
const ico_accuracy = require('../../images/icon_accuracy-24.png')
const ico_strength = require('../../images/icon_strength-24.png')
const ico_evasion = require('../../images/icon_evasion-32.png')
const ico_luck = require('../../images/icon_luck-24.png')
const ico_d10 = require('../../images/icon_d10-32.png')
const ico_insanity = require('../../images/ico_insanity-32.png')

// This is the individual survivor screen
@inject(({ store }, props) => ({
  survivor: store.selectedCampaign.settlement.survivors.get(props.survivorId),
}))
@observer
export default class Survivor extends React.Component {
  static propTypes = {
    survivorId: PropTypes.string.isRequired, // ID of the survivor to show
  }

  render() {
    let survivor = this.props.survivor

    if (!survivor) {
      return null
    }

    return (
      <View>
        <View styleName="horizontal">
          <Text>Name: {survivor.name}</Text>
          <GenderButton
            gender={survivor.gender}
            changeGender={survivor.changeGender}
          />

          <EditableProperty
            label="Survival"
            help="(Limit: 3)"
            minimumValue={0}
            maximumValue={3}
            showLabel={true}
            quantity={survivor.survival}
            setQuantity={qty => survivor.setAttribute('survival', qty)}
          />
        </View>
        <View styleName="horizontal">
          <EditableProperty
            label="Movement"
            icon={ico_movement}
            quantity={survivor.movement}
            setQuantity={qty => survivor.setAttribute('movement', qty)}
          />
          <EditableProperty
            label="Accuracy"
            icon={ico_accuracy}
            quantity={survivor.accuracy}
            setQuantity={qty => survivor.setAttribute('accuracy', qty)}
          />
          <EditableProperty
            label="Strength"
            icon={ico_strength}
            quantity={survivor.strength}
            setQuantity={qty => survivor.setAttribute('strength', qty)}
          />

          <EditableProperty
            label="Evasion"
            icon={ico_evasion}
            quantity={survivor.evasion}
            setQuantity={qty => survivor.setAttribute('evasion', qty)}
          />
          <EditableProperty
            label="Luck"
            icon={ico_luck}
            quantity={survivor.luck}
            setQuantity={qty => survivor.setAttribute('luck', qty)}
          />
          <EditableProperty
            label="Speed"
            icon={ico_d10}
            quantity={survivor.speed}
            setQuantity={qty => survivor.setAttribute('speed', qty)}
          />
          <EditableProperty
            label="Insanity"
            icon={ico_insanity}
            quantity={survivor.insanity}
            setQuantity={qty => survivor.setAttribute('insanity', qty)}
          />
        </View>

        <Divider />

        <View>
          <Subtitle>Courage: 3</Subtitle>
          <View styleName="horizontal">
            <Text>Stalwart</Text>
            <Text>Stalwart2</Text>
            <Text>Stalwart3</Text>
          </View>
        </View>

        <Divider />

        <View>
          <Subtitle>Understanding: 3</Subtitle>
          <View styleName="horizontal">
            <Text>Analyze</Text>
            <Text>Analyze2</Text>
            <Text>Analyze3</Text>
          </View>
        </View>

        <Divider />

        <FightingArts
          fightingArts={survivor.fightingArts}
          addFA={survivor.addFA}
          removeFA={survivor.removeFA}
        />

        <Divider />

        <Disorders
          disorders={survivor.disorders}
          addDisorder={survivor.addDisorder}
          removeDisorder={survivor.removeDisorder}
        />

        <Divider />

        <Note
          title="Notes"
          notes={survivor.notes}
          saveNote={survivor.saveNotes}
        />
      </View>
    )
  }
}
