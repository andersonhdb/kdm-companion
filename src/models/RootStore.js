import { types, getSnapshot } from 'mobx-state-tree'
import { SettlementLocation } from './SettlementLocation'
import { Campaign } from './Campaign'
import { Innovation } from './Innovation'
import { Resource } from './Resource'
import { Monster } from './Monster'
import { Gear } from './Gear'
import { Expansion } from './Expansion'
import { Principle } from './Principle'
import { FightingArt, SecretFightingArt } from './FightingArt'
import { Disorder } from './Disorder'
import { Ability } from './Ability'
import { WeaponProficiency } from './WeaponProficiency'
import { Subscription } from './Subscription'
import { uuid } from '../utils'
import R from 'ramda'

import locationsData from '../data/settlement_locations'
import innovationsData from '../data/innovations'
import fightingArtsData from '../data/fightingArts'
import secretFightingArtsData from '../data/secretFightingArts'
import disordersData from '../data/disorders'
import resourceData from '../data/resources'
import monsterData from '../data/monsters'
import gearData from '../data/gear'
import expansionData from '../data/expansions'
import principlesData from '../data/principles'
import abilitiesData from '../data/abilities'
import weaponProficiencyData from '../data/weaponProficiencies'

export default types
  .model('RootStore', {
    locations: types.optional(types.map(SettlementLocation), locationsData),
    innovations: types.optional(types.map(Innovation), innovationsData),
    fightingArts: types.optional(types.map(FightingArt), fightingArtsData),
    secretFightingArts: types.optional(
      types.map(SecretFightingArt),
      secretFightingArtsData
    ),
    disorders: types.optional(types.map(Disorder), disordersData),
    resources: types.optional(types.map(Resource), resourceData),
    monsters: types.optional(types.map(Monster), monsterData),
    gear: types.optional(types.map(Gear), gearData),
    expansions: types.optional(types.map(Expansion), expansionData),
    principles: types.optional(types.map(Principle), principlesData),
    abilities: types.optional(types.map(Ability), abilitiesData),
    weaponProficiencies: types.optional(
      types.map(WeaponProficiency),
      weaponProficiencyData
    ),
    // stuff stored to disc after this
    version: 1,
    campaigns: types.optional(types.array(Campaign), [
      {
        id: 'new',
      },
    ]),
    selectedCampaign: types.optional(types.reference(Campaign), 'new'),
    subscription: types.optional(Subscription, {}),
  })
  .actions(self => ({
    createCampaign(name) {
      id = uuid()
      self.campaigns.push({ id, settlement: { name: name, survivalLimit: 1 } })
      self.selectCampaign(id)
    },
    selectCampaign(id) {
      self.selectedCampaign = id
    },
    deleteCampaign(id) {
      self.campaigns = R.reject(c => c.id === id, self.campaigns)
      if (self.campaigns.length === 0) {
        self.createCampaign('New Settlement')
      } else {
        self.selectedCampaign = self.campaigns[0]
      }
    },
    setVersion(version) {
      self.version = version
    },
    load(data) {
      self.campaigns = data.campaigns
      self.selectedCampaign = data.selectedCampaign
      self.version = data.version || 1
      self.subscription = data.subscription || {}
    },
  }))
  .views(self => ({
    get data() {
      return {
        campaigns: getSnapshot(self.campaigns),
        selectedCampaign: self.selectedCampaign.id,
        version: self.version,
        subscription: getSnapshot(self.subscription),
      }
    },
    get campaignExpansions() {
      return self.selectedCampaign.expansionList
    },
    selectedExpansionFilter(map) {
      return self.selectedCampaign.selectedExpansionFilter(map)
    },
    get availableLocations() {
      return R.sortBy(
        R.prop('name'),
        self.selectedExpansionFilter(self.locations)
      )
    },
    get availableInnovations() {
      return R.sortBy(
        R.prop('name'),
        self.selectedExpansionFilter(self.innovations)
      )
    },
    get availableFightingArts() {
      return R.sortBy(
        R.prop('name'),
        self.selectedExpansionFilter(self.fightingArts)
      )
    },
    get availableSecretFightingArts() {
      return R.sortBy(
        R.prop('name'),
        self.selectedExpansionFilter(self.secretFightingArts)
      )
    },
    get availableDisorders() {
      return R.sortBy(
        R.prop('name'),
        self.selectedExpansionFilter(self.disorders)
      )
    },
    get availableAbilities() {
      return R.sortBy(
        R.prop('name'),
        self.selectedExpansionFilter(self.abilities)
      )
    },
    get availableHunts() {
      return self.selectedExpansionFilter(self.monsters)
    },
    get availableWeaponProficiencies() {
      return self.selectedExpansionFilter(self.weaponProficiencies)
    },
    get numCampaigns() {
      return self.campaigns.length
    },
  }))
