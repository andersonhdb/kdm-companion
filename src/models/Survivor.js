import { types } from 'mobx-state-tree'
import { FightingArt } from './FightingArt'
import { Disorder } from './Disorder'
import { uuid } from '../utils'
import R from 'ramda'

const Survivor = types
  .model('Survivor', {
    id: types.identifier,
    name: types.string,
    fightingArts: types.array(types.reference(FightingArt)),
    disorders: types.array(types.reference(Disorder)),
    gender: types.optional(types.enumeration(['male', 'female']), 'male'),
    survival: 1,
    movement: 5,
    accuracy: 0,
    strength: 0,
    evasion: 0,
    luck: 0,
    speed: 0,
    insanity: 0,
    notes: '',
  })
  .actions(self => ({
    addFA(fa) {
      let found = R.find(item => fa.id === item.id, self.fightingArts)
      if (!found) {
        self.fightingArts.push(fa)
      }
    },
    removeFA(fa) {
      // TODO: removing an element other than the last gives an error.
      self.fightingArts.remove(fa)
    },
    addDisorder(disorder) {
      let found = R.find(item => disorder.id === item.id, self.disorders)
      if (!found) {
        self.disorders.push(disorder)
      }
    },
    removeDisorder(disorder) {
      // TODO: removing an element other than the last gives an error.
      self.disorders.remove(disorder)
    },
    changeGender() {
      let gender = self.gender === 'male' ? 'female' : 'male'
      self.gender = gender
    },
    setAttribute(attribute, quantity) {
      if (self[attribute] !== undefined) {
        self[attribute] = quantity
      }
    },
    saveNotes(notes) {
      self.notes = notes
    },
  }))
const init = () => ({ id: uuid(), name: 'Unnamed', fightingArts: [] })

export { Survivor, init }
