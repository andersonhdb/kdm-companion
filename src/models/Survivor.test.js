import { getSnapshot } from 'mobx-state-tree'
import { observable } from 'mobx'
import { Survivor } from './Survivor'
import RootStore from './RootStore'

it('creates a survivor', () => {
  const survivor = Survivor.create({ id: 'test' })

  expect(getSnapshot(survivor)).toMatchSnapshot()
})

it('handles passed in properties', () => {
  const survivor = Survivor.create({ name: 'test', survival: 3, strength: 1 })

  expect(survivor).toHaveProperty('name', 'test')
  expect(survivor).toHaveProperty('survival', 3)
  expect(survivor).toHaveProperty('strength', 1)
})

it('.changeGender', () => {
  const survivor = Survivor.create({ id: 'test' })

  expect(survivor.gender).toBe('male')
  survivor.changeGender()
  expect(survivor.gender).toBe('female')
  survivor.changeGender()
  expect(survivor.gender).toBe('male')
})

it('cycleStatus', () => {
  const survivor = Survivor.create({ id: 'test' })

  expect(survivor.status).toBe('alive')
  survivor.cycleStatus()
  expect(survivor.status).toBe('dead')
  survivor.cycleStatus()
  expect(survivor.status).toBe('retired')
  survivor.cycleStatus()
  expect(survivor.status).toBe('alive')
})

describe('with RootStore', () => {
  let store
  let survivor

  beforeEach(() => {
    store = RootStore.create()
    survivor = store.selectedCampaign.settlement.createSurvivor('survivor')
  })

  it('has a name', () => {
    expect(survivor.name).toMatch('survivor')
  })
  it('handles fighting arts', () => {
    survivor.addFA({ id: 'Acrobatics' })
    survivor.addFA({ id: 'Clutch Fighter' })
    survivor.addFA({ id: 'Clarity of Darkness' })
    expect(survivor.fightingArts).toHaveLength(3)
    survivor.removeFA({ id: 'Clutch Fighter' })
    expect(survivor.fightingArts).toHaveLength(2)
  })

  it('handles disorders arts', () => {
    survivor.addDisorder({ id: 'Absent Seizures' })
    survivor.addDisorder({ id: 'Fear of the Dark' })
    survivor.addDisorder({ id: 'Indecision' })
    expect(survivor.disorders).toHaveLength(3)
    survivor.removeDisorder({ id: 'Fear of the Dark' })
    expect(survivor.disorders).toHaveLength(2)
  })

  it('handles abilities', () => {
    survivor.addAbility({ id: 'Acid Palms' })
    survivor.addAbility({ id: 'Homing Instinct' })
    survivor.addAbility({ id: 'Metal Maw' })
    survivor.addAbility({ id: 'Way of the Rust' })
    expect(survivor.abilities).toHaveLength(4)
    survivor.removeAbility({ id: 'Metal Maw' })
    expect(survivor.abilities).toHaveLength(3)
  })

  it('.setWeaponProficiency', () => {
    survivor.setWeaponProficiency({ id: 'fist & tooth' })
    expect(survivor.weaponProficiency.id).toBe('fist & tooth')
  })

  it('has initial survival', () => {
    expect(survivor.survival).toBe(1)
  })

  it('gives all bonus', () => {
    store.selectedCampaign.setCampaignType('pots')
    store.selectedCampaign.selectInnovation({ id: 'saga' }) // courage: 2, understanding: 2, 'hunt xp': 2,
    store.selectedCampaign.selectInnovation({ id: 'clan_of_death' }) //  accuracy: 1, strength: 1, evasion: 1,
    store.selectedCampaign.selectInnovation({ id: 'radiating orb' }) //  survival: 1,
    store.selectedCampaign.selectInnovation({ id: 'empire' }) //  strength: 1,

    store.selectedCampaign.selectPrinciple('newlife', { id: 'sotf' }) // strength: 1, evasion: 1,
    store.selectedCampaign.selectPrinciple('death', { id: 'graves' }) // understanding: 1,
    store.selectedCampaign.selectPrinciple('conviction', { id: 'barbaric' }) // strength: 1,

    survivor2 = store.selectedCampaign.settlement.createSurvivor('survivor2')

    expect(survivor2).toMatchObject({
      strength: 4,
      evasion: 2,
      accuracy: 1,
      courage: 2,
      'hunt xp': 2,
      insanity: 0,
      luck: 0,
      movement: 5,
      speed: 0,
      survival: 2,
      understanding: 3,
    })

    expect(survivor2.dragonTraits).toHaveProperty('accuracy', true)
    expect(survivor2.dragonTraits).toHaveProperty('strength', true)
  })
})
