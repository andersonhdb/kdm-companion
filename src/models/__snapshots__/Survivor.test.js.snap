// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates a survivor 1`] = `
Object {
  "abilities": Array [],
  "accuracy": 0,
  "cannotUseAbilities": false,
  "cannotUseFightingArts": false,
  "cannotUseSurvival": false,
  "courage": 0,
  "disorders": Array [],
  "dragonTraits": Object {
    "accuracy": false,
    "championsRite": false,
    "courage": false,
    "destined": false,
    "fatedBlow": false,
    "frozenStar": false,
    "iridescentHide": false,
    "noble": false,
    "oraclesEye": false,
    "pristine": false,
    "reincarnated": false,
    "scar": false,
    "strength": false,
    "unbreakable": false,
    "understanding": false,
    "weaponMaster": false,
  },
  "evasion": 0,
  "fightingArts": Array [],
  "gender": "male",
  "hunt xp": 0,
  "id": "test",
  "insanity": 0,
  "luck": 0,
  "movement": 5,
  "name": "Unnamed",
  "notes": "",
  "rerollUsed": false,
  "skipNextHunt": false,
  "speed": 0,
  "status": "alive",
  "strength": 0,
  "survival": 1,
  "understanding": 0,
  "weaponProficiency": undefined,
  "weaponProficiencyLevel": 0,
}
`;
